python/lib/sift_py/upload/config.py

```diff
@@ -0,0 +1,149 @@
+from typing import Dict, List, Optional
+
+from pydantic import BaseModel, ConfigDict, model_validator
+from pydantic_core import PydanticCustomError
+
+VALID_TIME_FORMATS = [
+    "TIME_FORMAT_ABSOLUTE_RFC3339",
+    "TIME_FORMAT_ABSOLUTE_DATETIME",
+    "TIME_FORMAT_ABSOLUTE_UNIX_SECONDS",
+    "TIME_FORMAT_ABSOLUTE_UNIX_MILLISECONDS",
+    "TIME_FORMAT_ABSOLUTE_UNIX_MICROSECONDS",
+    "TIME_FORMAT_ABSOLUTE_UNIX_NANOSECONDS",
+    "TIME_FORMAT_RELATIVE_NANOSECONDS",
+    "TIME_FORMAT_RELATIVE_MICROSECONDS",
+    "TIME_FORMAT_RELATIVE_MILLISECONDS",
+    "TIME_FORMAT_RELATIVE_SECONDS",
+    "TIME_FORMAT_RELATIVE_MINUTES",
+    "TIME_FORMAT_RELATIVE_HOURS",
+]
+
+
+VALID_DATA_TYPES = [
+    "CHANNEL_DATA_TYPE_DOUBLE",
+    "CHANNEL_DATA_TYPE_FLOAT",
+    "CHANNEL_DATA_TYPE_STRING",
+    "CHANNEL_DATA_TYPE_BOOL",
+    "CHANNEL_DATA_TYPE_INT_32",
+    "CHANNEL_DATA_TYPE_INT_64",
+    "CHANNEL_DATA_TYPE_UINT_32",
+    "CHANNEL_DATA_TYPE_UINT_6",
+    "CHANNEL_DATA_TYPE_ENUM",
+    "CHANNEL_DATA_TYPE_BIT_FIELD",
+]
+
+
+class _BaseModel(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+
+class EnumType(_BaseModel):
+    key: int
+    name: str
+
+
+class BitFieldElement(_BaseModel):
+    index: int
+    name: str
+    bit_count: int
+
+
+class TimeColumn(_BaseModel):
+    format: str
+    column_number: int
+    relative_start_time: Optional[str] = None
+
+    @model_validator(mode="after")
+    def validate_format(self):
+        if self.format not in VALID_TIME_FORMATS:
+            raise PydanticCustomError(
+                "invalid_config_error",
+                "Invalid time format: {format}.\nValid options: {valid}",
+                {"format": self.format, "valid": ", ".join(VALID_TIME_FORMATS)},
+            )
+
+        return self
+
+    @model_validator(mode="after")
+    def validate_relative_time(self):
+        if self.format.startswith("TIME_FORMAT_RELATIVE_"):
+            if self.relative_start_time is None:
+                raise PydanticCustomError("invalid_config_error", "Missing 'relative_start_time'")
+        else:
+            if self.relative_start_time:
+                raise PydanticCustomError(
+                    "invalid_config_error",
+                    "'relative_start_time' specified for non relative time format.",
+                )
+
+        return self
+
+
+class DataColumn(_BaseModel):
+    name: str
+    data_type: str
+    component: Optional[str] = ""
```

Comment:
Having an optional default to a blank string (falsey) feels a little awkward and redundant. If we're going to default to non-`None` values could we get rid of the optional?